### PR TITLE
Basic Atomic for NuttX

### PR DESCRIPTION
**Pull request by crafcat7**
## Summary
Enhanced NuttX support for stdatomic

1. Added stdatomic for NuttX. When the compiler version is lower than C11/C++17 and Atomic functions are needed, the NuttX stdatomic version will provide atomic support. This is a wrapper of the atomic function implemented in arch_atomic.c, and uses spinlock to ensure atomic operations.
2. In include/nuttx provides atomic.h.This is the header file used to distribute atomic logic. When our toolchain supports atomic, we should give priority to the version provided by the toolchain. When it is not supported, we will use the NuttX version of stdatomic.
3. Replaced the use of stdatomic.h in the NuttX directory and pointed it to nuttx/atomic.h

## Impact
Added stdatomic.h in NuttX, and changed all places using #include <stdatomic.h> to #include <nuttx/atomic.h>, to ensure that when the toolchain does not support stdatomic.h capabilities, the NuttX version is used.

## Testing
Local Test Pass

### CODE REVIEW

File: libs/libc/machine/CMakeLists.txt [hunk 1]
```diff
@@ -20,6 +20,6 @@
 
 add_subdirectory(${CONFIG_ARCH})
 
-if(CONFIG_LIBC_ARCH_ATOMIC)
-  target_sources(c PRIVATE arch_atomic.c)
+target_sources(c PRIVATE arch_atomic.c)
+
 endif()
```
Comment (xuxin930) [inline libs/libc/machine/CMakeLists.txt:25]: does this endif() should be deleted ？

Comment (crafcat7) [reply libs/libc/machine/CMakeLists.txt]: Done

**Review by acassis — APPROVED** [review]

**Comment by PetervdPerk-NXP** [comment]
I think it would be a good idea to add an example/test in nuttx/apps using libcxxmini.
So CI will test the implementation with different toolchains.

**Comment by xiaoxiang781216** [comment]
> I think it would be a good idea to add an example/test in nuttx/apps using libcxxmini. So CI will test the implementation with different toolchains.

exactly, we already write one. @crafcat7 please upstream the test code.

**Comment by crafcat7** [comment]
> I think it would be a good idea to add an example/test in nuttx/apps using libcxxmini. So CI will test the implementation with different toolchains.

We can use this atomic test in
https://github.com/apache/nuttx-apps/blob/master/testing/atomic/atomic_main.c

**Comment by PetervdPerk-NXP** [comment]
> We can use this atomic test in
https://github.com/apache/nuttx-apps/blob/master/testing/atomic/atomic_main.c

Looks good to me as long we're sure every architecture is triggering this on the CI. Atomics and toolchains are a bit of a pain.

**Comment by xiaoxiang781216** [comment]
> > We can use this atomic test in
> > https://github.com/apache/nuttx-apps/blob/master/testing/atomic/atomic_main.c
> 
> Looks good to me as long we're sure every architecture is triggering this on the CI. Atomics and toolchains are a bit of a pain.

in the next patchset, `struct file` will add atomic reference count which should cover all active maintained arch.

**Pull request merged**